### ISSUE TITLE
A Visual And Interactive Look at Basic Neural Network Math

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#6947]
A Visual And Interactive Look at Basic Neural Network Math<br>
<br>
https://ift.tt/2INW1SQ<br>
<br>
Jay Alammar<br>
<br>
<div>
<div>
<div>
<div>
<p>In the <a href="https://jalammar.github.io/visual-interactive-guide-basics-neural-networks/">previous post, we looked at the basic concepts of neural networks</a>. Let us now take another example as an excuse to guide us to explore some of the basic mathematical ideas involved in prediction with neural networks.</p>
<video> Your browser does not support the video tag.</video>
<p>If you had been aboard the Titanic, would you have survived the sinking event? Let&rsquo;s build a model to predict one&rsquo;s odds of survival.</p>
<p>This will be a neural network model building on what we discussed in the <a href="https://jalammar.github.io/feedforward-neural-networks-visual-interactive/">previous post</a>, but will have a higher prediction accuracy because it utilizes hidden layers and activation functions.</p>
<p>The dataset we&rsquo;ll use this time will be the <a href="https://jalammar.github.io/feedforward-neural-networks-visual-interactive/">Titanic passenger list</a> from Kaggle. It lists the names and other information of the passengers and shows whether each passenger survived the sinking event or not.</p>
<p>The raw dataset looks like this:</p>
<div>
<table>
<thead>
<tr>
<th>PassengerId</th>
<th>Survived</th>
<th>Pclass</th>
<th>Name</th>
<th>Sex</th>
<th>Age</th>
<th>SibSp</th>
<th>Parch</th>
<th>Ticket</th>
<th>Fare</th>
<th>Cabin</th>
<th>Embarked</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>0</td>
<td>3</td>
<td>Braund, Mr. Owen Harris</td>
<td>male</td>
<td>22.0</td>
<td>1</td>
<td>0</td>
<td>A/5 21171</td>
<td>7.2500</td>
<td>NaN</td>
<td>S</td>
</tr>
<tr>
<td>2</td>
<td>1</td>
<td>1</td>
<td>Cumings, Mrs. John Bradley (Florence Briggs Th&hellip;</td>
<td>female</td>
<td>38.0</td>
<td>1</td>
<td>0</td>
<td>PC 17599</td>
<td>71.2833</td>
<td>C85</td>
<td>C</td>
</tr>
<tr>
<td>3</td>
<td>1</td>
<td>3</td>
<td>Heikkinen, Miss. Laina</td>
<td>female</td>
<td>26.0</td>
<td>0</td>
<td>0</td>
<td>STON/O2. 3101282</td>
<td>7.9250</td>
<td>NaN</td>
<td>S</td>
</tr>
</tbody>
</table>
</div>
<p>We won&rsquo;t bother with most of the columns for now. We&rsquo;ll just use the sex and age columns as our features, and survival as our label that we&rsquo;ll try to predict.</p>
<div>
<table>
<thead>
<tr>
<th>Age</th>
<th>Sex</th>
<th>Survived?</th>
</tr>
</thead>
<tbody>
<tr>
<td>22</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>38</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>26</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td colspan="3">&hellip; 891 rows total</td>
</tr>
</tbody>
</table>
</div>
<p>We&rsquo;ll attempt to build a network that predicts whether a passenger survived or not.</p>
<p>Neural networks need their inputs to be numeric. So we had to change the sex column &ndash; male is now 0, female is 1. You&rsquo;ll notice the dataset already uses something similar for the survival column &ndash; survived is 1, did not survive is 0.</p>
<p>The simplest neural network we can use to train to make this prediction looks like this:</p>
<div>
<img alt="neural netowrk with two inputs, one output, and sigmoid output activation" src="https://jalammar.github.io/images/two-input-one-output-sigmoid-network.png">
<p>Calculating a prediction is done by plugging in a value for "age" and "sex". The calculation then flows from the left to the right. Before we can use this net for prediction, however, we'll have to run a "training" process that will give us the values for the weights (</p>
<span>w</span>
<p>) and bias (</p>
<span>b</span>
<p>).</p>
<p>Note: we have slightly adjusted the way we represent the networks from the previous post. The bias node specifically is more commonly represented like this</p>
</div>
<p>Let&rsquo;s recap the elements that make up this network and how they work:</p>
<div>
<div>
<p>An input neuron is where we plug in an input value (e.g. the age of a person). It&rsquo;s where the calculation starts. The outgoing connection and the rest of the graph tell us what other calculations we need to do to calculate a prediction.</p>
</div>
</div>
<div>
<div>
<p>If a connection has a weight, then the value is multiplied by that weight as it passes through it.</p>
<div>
<div>
<pre><code>connection_output = weight * connection_input
</code></pre>
</div>
</div>
</div>
</div>
<div>
<div>
<p>If a neuron has inputs, it sums their value and sends that sum along its outgoing connection(s).</p>
<div>
<div>
<pre><code>node_output = input_1 + input_2
</code></pre>
</div>
</div>
</div>
</div>
<h3>Sigmoid <a name="sigmoid" href="https://jalammar.github.io/feedforward-neural-networks-visual-interactive/#sigmoid" id="sigmoid">#</a>
</h3>
<div>
<div>
<p>To turn the network&rsquo;s calculation into a probability value between 0 and 1, we have to pass the value from the output layer through a &ldquo;sigmoid&rdquo; formula. Sigmoid squashes the output value of a neuron to between 0 and 1 according to a specific curve.</p>
<p>f(x)=</p>
<p>Where e is the mathematical constant approximately equal to 2.71828</p>
<div>
<div>
<pre><code>def sigmoid(x):
    return 1/(1 + math.exp(-x))

output = sigmoid(value)
</code></pre>
</div>
</div>
</div>
</div>
<h3>Sigmoid Visualization <a name="sigmoid-visualization" href="https://jalammar.github.io/feedforward-neural-networks-visual-interactive/#sigmoid-visualization" id="sigmoid-visualization">#</a>
</h3>
<div>
<div>
<table>
<tr>
<td valign="middle">0</td>
<td><img src="https://jalammar.github.io/images/sigmoid-neuron.png"></td>
<td></td>
</tr>
</table>
</div>
<div>
<p>Interact a little with sigmoid to see how it transforms various values</p>
<p>f(<span><span>0</span></span>) = <span><span><span><span><span><span><span><span><span><span>&#8203;</span></span><span><span><span>1</span><span>+</span> <span><span>e</span> <span><span><span><span>&#8203;</span></span> <span><span><span>&minus;(<span>x</span>)</span></span></span></span> <span><span><span>&#8203;</span></span>&#8203;</span></span></span></span></span></span><span><span><span>&#8203;</span></span></span><span><span><span>&#8203;</span></span><span><span><span>1</span></span></span></span><span><span><span>&#8203;</span></span>&#8203;</span></span></span></span></span></span></span></span> =</p>
</div>
</div>
<div>
<div>
<p>To bring it all together, calculating a prediction with this shallow network looks like this:</p>
<div>
<div>
<pre><code>def sigmoid(x):
    return 1/(1 + math.exp(-x))

def calculate_prediction(age, sex, weight_1, weight_2, bias):

    # Multiply the inputs by their weights, sum the results up
    layer_2_node = age * weight_1 + sex * weight_2 + 1 * bias

    prediction = sigmoid(layer_2_node)
    return prediction
</code></pre>
</div>
</div>
</div>
</div>
<p>Now that we know the structure of our network, we can train it using [gradient descent] running on the first 600 rows of the 891-row dataset. I will not be addressing the training process in this post because that&rsquo;s a separate concern at the moment. For now, I just want you to be comfortable with how a trained network calculates a prediction. Once you get this intuition down, we&rsquo;ll proceed to training in a future post.</p>
<p>The training process gives us the following values (with an accuracy of 73.20%):</p>
<div>
<div>
<pre><code>weight_1 =   -0.016852 # Associated with "Age"
weight_2 =   0.704039  # Associated with "Sex" (where male is 0, female is 1)
bias =       -0.116309
</code></pre>
</div>
</div>
<p>Intuitively, the weights indicate how much their associated property contribute to the prediction &ndash; odds of survival improve the younger a person is (since a larger age multiplied by the negative weight value gives a bigger negative number). They improve more if the person is female.</p>
<h3>Prediction Calculation <a name="prediction-calculation" href="https://jalammar.github.io/feedforward-neural-networks-visual-interactive/#prediction-calculation" id="prediction-calculation">#</a>
</h3>
<p>The trained network now looks like this: (hover or click on values in the table to see their individual predictions)</p>
<p>An accuracy of 73.20% isn&rsquo;t very impressive. This is a case that can benefit from adding a hidden layer. Hidden layers give the model more capacity to represent more sophisticated prediction functions that may do a better job (<a href="http://www.deeplearningbook.org/contents/ml.html">Deep Learning ch.5 page 113</a>).</p>
<div>
<div>
<p>It&rsquo;s often useful to apply certain math functions to the weighted outputs. These are called &ldquo;activation functions&rdquo; because historically they translated the output of the neuron into either 1 (On/active) or 0 (Off).</p>
<div>
<div>
<pre><code>def activation_function(x):
    # Do something to the value
    ...

weighted_sum = weight * (input_1 + input_2)
output = activation_function(weighted_sum)
</code></pre>
</div>
</div>
<p>Activation functions are vital for hidden layers. Without them, deep networks would be no better than a shallow linear network. Read the &ldquo;Commonly used activation functions&rdquo; section from <a href="https://cs231n.github.io/neural-networks-1/">Neural Networks Part 1: Setting up the Architecture</a> for a look at various activation functions.</p>
</div>
</div>
<h3>ReLU <a name="relu" href="https://jalammar.github.io/feedforward-neural-networks-visual-interactive/#relu" id="relu">#</a>
</h3>
<div>
<div>
<p>A leading choice for activation function is called ReLU. It returns 0 if its input is negative, returns the number itself otherwise. Very simple!</p>
<p>f(x) = max(0, x)</p>
<div>
<div>
<pre><code># Naive scalar relu implementation. In the real world, most calculations are done on vectors
def relu(x):
    if x &lt; 0:
        return 0
    else:
        return x

output = relu(value)
</code></pre>
</div>
</div>
</div>
</div>
<h3>ReLU Visualization <a name="relu-visualization" href="https://jalammar.github.io/feedforward-neural-networks-visual-interactive/#relu-visualization" id="relu-visualization">#</a>
</h3>
<div>
<div>
<table>
<tr>
<td valign="middle">0</td>
<td><img src="https://jalammar.github.io/images/relu.png"></td>
<td></td>
</tr>
</table>
</div>
<div>
<p>Interact a little with relu to see how it transforms various values</p>
<p>f(<span><span>0</span></span>) = max( 0, <span><span>x</span></span>) =</p>
</div>
</div>
<h2>Closing</h2>
<p>This post has been parked for more than a year. I had attempted to visualize a deeper network after this point, but that never materialized. I hope you enjoyed it. Let me know on <a href="https://twitter.com/JayAlammar">@JayAlammar on Twitter</a> if you have any feedback.</p>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via GitHub Pages | Websites for you and your projects, hosted directly from your GitHub repository. Just edit, push, and your changes are live.<br>
<br>
July 28, 2020 at 04:39PM
